File: euro_aip/euro_aip/briefing/models/flight_exchange.py
"""FlightExchange — cross-app flight interchange model.

A shared, language-neutral representation of a "flight" (route + timing +
aircraft) that round-trips between the Python server (``euro_aip``) and the
Swift apps (``RZFlight``), so flights can be exported from one flyfun app and
imported into another.

``FlightExchange`` is a thin envelope around the existing :class:`Route` type:
the ``route`` JSON object is :meth:`Route.to_dict` embedded verbatim, so the
two never drift. The envelope only adds the fields that are *not* on ``Route``:
a display ``name``, the aircraft ``registration``, provenance (``source``), and
a ``schema_version``.

See ``designs/flight_exchange_design.md`` for the full wire format and field
semantics. The Swift counterpart lives in
``Sources/RZFlight/Briefing/FlightExchange.swift`` and must stay in parity.
"""

from dataclasses import dataclass
from typing import Optional

from euro_aip.briefing.models.route import Route

# Wire format version. v1 == designs/flight_exchange_design.md.
# Consumers reject unknown major versions.
SCHEMA_VERSION = 1


@dataclass
class FlightExchange:
    """Cross-app flight interchange envelope around a :class:`Route`.

    Attributes:
        route: The flight route. Serialized verbatim as ``Route.to_dict()`` —
            the single source of route truth (departure/destination/waypoints/
            alternates, coords, times, cruise altitude, aircraft type).
        schema_version: Wire format version (v1 = current).
        name: Optional display title (e.g. "Oxford -> Sion").
        aircraft_registration: Aircraft registration — the cross-app field that
            no app stores at the route level today.
        source_app: Provenance — emitting app ("weather" | "forms" | "brief").
        source_flight_id: Sender's native flight id (opaque to the consumer).
        source_share_code: Present when the sender supports public re-fetch.
    """

    route: Route
    schema_version: int = SCHEMA_VERSION
    name: Optional[str] = None
    aircraft_registration: Optional[str] = None
    source_app: Optional[str] = None
    source_flight_id: Optional[str] = None
    source_share_code: Optional[str] = None

    def to_dict(self) -> dict:
        """Serialize to the language-neutral wire format (snake_case keys).

        Optional envelope fields are omitted when unset, so a minimal flight
        emits just ``schema_version`` and ``route``. The ``aircraft.type`` key
        is a convenience mirror of ``route.aircraft_type`` (the route remains
        the source of truth).
        """
        data: dict = {
            'schema_version': self.schema_version,
            'route': self.route.to_dict(),
        }

        if self.name is not None:
            data['name'] = self.name

        source = {}
        if self.source_app is not None:
            source['app'] = self.source_app
        if self.source_flight_id is not None:
            source['flight_id'] = self.source_flight_id
        if self.source_share_code is not None:
            source['share_code'] = self.source_share_code
        if source:
            data['source'] = source

        aircraft = {}
        if self.aircraft_registration is not None:
            aircraft['registration'] = self.aircraft_registration
        # Convenience mirror of route.aircraft_type.
        if self.route.aircraft_type is not None:
            aircraft['type'] = self.route.aircraft_type
        if aircraft:
            data['aircraft'] = aircraft

        return data

    @classmethod
    def from_dict(cls, data: dict) -> 'FlightExchange':
        """Create from the wire format.

        ``aircraft.type`` is ignored on import — ``route.aircraft_type`` is the
        source of truth, and ``to_dict`` re-mirrors it on the way out.
        """
        # Reject unknown (newer) schema versions before touching any fields —
        # a v2 payload may carry breaking changes this build can't interpret.
        # See design doc.
        version = data.get('schema_version', SCHEMA_VERSION)
        if version > SCHEMA_VERSION:
            raise ValueError(
                f"Unsupported FlightExchange schema_version {version} "
                f"(max supported {SCHEMA_VERSION})"
            )

        source = data.get('source') or {}
        aircraft = data.get('aircraft') or {}

        return cls(
            route=Route.from_dict(data['route']),
            schema_version=version,
            name=data.get('name'),
            aircraft_registration=aircraft.get('registration'),
            source_app=source.get('app'),
            source_flight_id=source.get('flight_id'),
            source_share_code=source.get('share_code'),
        )

    def __repr__(self) -> str:
        return (
            f"FlightExchange({self.route.departure} -> {self.route.destination}"
            f", v{self.schema_version})"
        )

File: Sources/RZFlight/Briefing/FlightExchange.swift
//
//  FlightExchange.swift
//  RZFlight
//
//  Cross-app flight interchange model.
//  Mirrors the Python euro_aip.briefing.models.FlightExchange wire format.
//

import Foundation

/// A shared, language-neutral representation of a "flight" (route + timing +
/// aircraft) that round-trips between the Python server (`euro_aip`) and the
/// Swift apps (`RZFlight`), so a flight can be exported from one flyfun app and
/// imported into another.
///
/// `FlightExchange` is a thin envelope around the existing ``Route`` type: the
/// `route` JSON object is `Route`'s coding embedded verbatim, so the two never
/// drift. The envelope only adds the fields that are *not* on ``Route``: a
/// display ``name``, the aircraft ``registration``, provenance (``source``),
/// and a ``schemaVersion``.
///
/// See `designs/flight_exchange_design.md` for the full wire format. The Python
/// counterpart lives in `euro_aip/briefing/models/flight_exchange.py` and must
/// stay in parity.
public struct FlightExchange: Codable, Sendable {

    /// Wire format version. v1 == designs/flight_exchange_design.md.
    public static let currentSchemaVersion: Int = 1

    // MARK: - Nested Types

    /// Provenance — where the flight came from. All fields optional.
    public struct Source: Codable, Sendable {
        /// Emitting app: "weather" | "forms" | "brief".
        public var app: String?

        /// Sender's native flight id (opaque to the consumer).
        public var flightId: String?

        /// Present when the sender supports public re-fetch.
        public var shareCode: String?

        enum CodingKeys: String, CodingKey {
            case app
            case flightId = "flight_id"
            case shareCode = "share_code"
        }

        public init(app: String? = nil, flightId: String? = nil, shareCode: String? = nil) {
            self.app = app
            self.flightId = flightId
            self.shareCode = shareCode
        }
    }

    /// Aircraft envelope — the cross-app fields not carried on ``Route``.
    public struct Aircraft: Codable, Sendable {
        /// Aircraft registration — the cross-app field none store on the route today.
        public var registration: String?

        /// Convenience mirror of `route.aircraftType` (the route stays the source of truth).
        public var type: String?

        public init(registration: String? = nil, type: String? = nil) {
            self.registration = registration
            self.type = type
        }
    }

    // MARK: - Fields

    /// Wire format version (v1 = current).
    public var schemaVersion: Int

    /// Optional provenance.
    public var source: Source?

    /// Optional display title (e.g. "Oxford -> Sion").
    public var name: String?

    /// The flight route — the single source of route truth.
    public var route: Route

    /// Optional aircraft envelope (registration + type mirror).
    public var aircraft: Aircraft?

    // MARK: - CodingKeys

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case source
        case name
        case route
        case aircraft
    }

    // MARK: - Decoding

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.schemaVersion = try container.decodeIfPresent(Int.self, forKey: .schemaVersion)
            ?? FlightExchange.currentSchemaVersion
        // Reject unknown (newer) schema versions — a v2 payload may carry
        // breaking changes this build can't interpret. See design doc.
        guard self.schemaVersion <= FlightExchange.currentSchemaVersion else {
            throw DecodingError.dataCorruptedError(
                forKey: .schemaVersion, in: container,
                debugDescription: "Unsupported FlightExchange schema_version "
                    + "\(self.schemaVersion) (max supported "
                    + "\(FlightExchange.currentSchemaVersion))")
        }
        self.source = try container.decodeIfPresent(Source.self, forKey: .source)
        self.name = try container.decodeIfPresent(String.self, forKey: .name)
        self.route = try container.decode(Route.self, forKey: .route)
        self.aircraft = try container.decodeIfPresent(Aircraft.self, forKey: .aircraft)
    }
}

// MARK: - Memberwise Initializer

extension FlightExchange {
    /// Create a FlightExchange with all fields specified.
    public init(
        route: Route,
        schemaVersion: Int = FlightExchange.currentSchemaVersion,
        name: String? = nil,
        source: Source? = nil,
        aircraft: Aircraft? = nil
    ) {
        self.route = route
        self.schemaVersion = schemaVersion
        self.name = name
        self.source = source
        self.aircraft = aircraft
    }
}

// MARK: - Convenience Coding

extension FlightExchange {
    /// Decode a FlightExchange from JSON data, using the snake_case + ISO-8601
    /// conventions of the Python wire format.
    public static func decode(from data: Data) throws -> FlightExchange {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(FlightExchange.self, from: data)
    }

    /// Encode this FlightExchange to JSON data in the language-neutral wire
    /// format (snake_case keys, ISO-8601 UTC times).
    public func encoded() throws -> Data {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return try encoder.encode(self)
    }
}

// MARK: - Protocol Conformances

extension FlightExchange: CustomStringConvertible {
    public var description: String {
        return "FlightExchange(\(route.departure) -> \(route.destination), v\(schemaVersion))"
    }
}
